### DISCUSSION TITLE
Idea: convert all "issues" that are requesting a feature into "discussion" under feature-request category

### DISCUSSION THREAD

**Discussion by keradus** [PHP-CS-Fixer/PHP-CS-Fixer#7498] [RFC]
That would:
- allow to easily vote (and sort by voting)
- allow to easily separate bugs ("issues") and feature requests ("discussions", [feature ideas](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/discussions/categories/feature-ideas) category)

**Comment by Wirone** [comment]
I am against moving _all_ issues. IMHO the idea of discussions is to have an initial idea and then prepare the final output - specification for implementation. Some of the feature requests are totally valid and moving them to discussions would mean that we no longer consider them as something to do.

From my perspective the distinction should be like this:
- keep feature requests as issues if we consider implementing them as a core team (when it can happen is a separate thing). Then we can track them easily, stale bot can help us with maintenance and ping us.
- keep feature requests as discussions if we consider accepting it as an external contribution. These should be guided, accepted and awaiting for someone to implement 🙂. That does not mean the core team can't pick it up.

Extended explanation [here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/discussions/7498#discussioncomment-7762084).

---

In terms of reasoning from the RFC:

> allow to easily vote (and sort by voting)

Issues allow the same, you can [sort issues by 👍 reaction](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

> allow to easily separate bugs ("issues") and feature requests ("discussions", [feature ideas](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/discussions/categories/feature-ideas) category)

["Issues" ≠ "Bugs"](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/discussions/7498#discussioncomment-7760873)

**Reply by keradus** [reply]
I like the proposal, and with following it - i suggest to convert 95% of valid feature requests, as core team won't be focusing on them (we rather focus nowadays on compatibility, 8.x, CI, etc).
To simplify decision "what to move", i would move 100% and for 5% core team is committed to deliver, we can rely on your 2nd bullet point ;)

**Reply by keradus** [reply]
now, we have 99 feature requests under "issues"

**Reply by Wirone** [reply]
I already moved some of the issues when reviewing the backlog. What was left in the issues I personally consider as something worth implementing, but my priority are PRs - I wanted to finish them, then start to work on issues.

**Reply by Wirone** [reply]
> now, we have 99 feature requests under "issues"

The "issue" term does not have negative meaning on Github, it does not represent bugs only. It's rather like "matter", an application. That's why feature requests can live under issues, IMHO the only requirement is them to have proper definition of done.

Of course, moving them to Discussions would lower the amount of maintenance (stale bot), but on the other hand it would be harder to keep up with them and auto-close them when they're implemented (PRs don't link to discussions).

**Reply by keradus** [reply]
> keep feature requests as issues if we consider implementing them as a core team

Do we consider implementing those 99 requests as corr team? No. If we don't plan to (but we are ok if someone does), we can move them, right ?

**Reply by Wirone** [reply]
> Do we consider implementing those 99 requests as core team? No.

Why are you speaking in the name of every member 🤔? Personally I see many of these feature requests as interesting and worth implementing. Really, I already spent a lot of time on triaging and made that consideration before. What I meant before is "issues that can be picked up by core team (or any contributor) at an unspecified time" should stay in issues, while feature requests that "100% won't be picked up by core team, but can be merged if anyone is willing to work on it after specification is done" should stay in "Feature ideas" section of Discussions.

That would make clear contract between core team and the contributors - if feature request is under Discussions, then it's required to get a green light before implementing it, but if it's under Issues, then I can work on it and it will be processed.

Having said that, we should move only issues:
- without clear specifications and definition of done
- only if it can't be inferred from existing comments (for clarity the summary from comments could be moved to issue's description, if it's not there already)
- only if we really, really don't consider working on them by ourselves

Feature requests that are under Issues are in line with this approach, because I left only those that I personally agree with and would approve proper implementation. I am not sure if spec/DoD is in the each one's description, though.

**Reply by keradus** [reply]
eg https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/7500 , i am not conviced it fits any of category to be in "issues", and if anyone (core or not) want to implement it, we like to see it :+1:

> Why are you speaking in the name of every member 🤔? 

you plan all of them? great to hear! is it realistic to expect you/we deliver on it in reasonable time? i really doubt. it's nothing wrong, but if you say "we commit to those 100 ideas", it is not that we really committed to any of them.

I would love to have a single place with all feature requests, either "accepted, community can bring implementation" (and if core also want to, well, we are the community too!) or "rejected, closed".

Now, we have 2 places, no clear distinguishes for me, I'm getting lost what is where and ignoring some of the topics because I do not recall to go to both places.

**Reply by Wirone** [reply]
> you plan all of them? great to hear! is it realistic to expect you/we deliver on it in reasonable time? i really doubt. it's nothing wrong, but if you say "we commit to those 100 ideas", it is not that we really committed to any of them.

Moving everything to Discussions does not solve this, it's just a brushing everything under the carpet and avoiding responsibility IMHO. I believe that everyone is aware that it's not like we spend all of our free time on implementing stuff here. But if the issue is there, it can be picked up at some point. I don't commit to all of them, but as I already said I would resolve some of them after we clean up opened PRs (some of the issues most probably _will_ be resolved with these PRs).

Yes, these feature requests can be picked up also from Discussions' feature ideas sections, but as I said - it's worse when it comes to automation, since discussions can't be linked to pull requests.

> I would love to have a single place with all feature requests, either "accepted, community can bring implementation" (and if core also want to, well, we are the community too!) or "rejected, closed".
> 
> Now, we have 2 places, no clear distinguishes for me, I'm getting lost what is where and ignoring some of the topics because I do not recall to go to both places.

I don't want to be rude, but when I joined Fixer there were 300+ issues and you did not seem interested in triaging them in any way. We were able to get rid of more than 50% of them, and I've put a lot of work to make it happen. Now, I am triaging any new issues and I react to Stale Bot's activity. You don't need to feel lost, because in most cases I handle it. I left many feature requests under Issues section on purpose - to be able to keep track on them. If we move all of them to Discussions, automation will be lost for these feature request, because Stale Bot does not cover Discussions. Effectively it means no one will look at them cyclically to decide if they are still relevant, like it's now.

Anyway, the poll shows that I am minority here, so maybe there's no point in discussing it anymore. You do you, you have final word here. I described how I see it, but you're free to do as you wish, I will adapt 🙂.

Regardless of the decision, this must be documented in the contributing guide (or other place), so there's a clear path for bugs, technical requirements (like new language features), feature ideas etc.

**Reply by keradus** [reply]
I recognise the argument of PR linking discussion or stale bot. It is a big change in perception on my side, wasn't aware of this part.

For stale bot, I believe that we can allow stale bot to _close_ issues without your manual supervision. If person who was interested in topic will be still interested in it, they can always ask to reopen or... wait for it... create a PR.
and if we have idea of long-term change, we park it under dedicated milestone.

i like idea of documenting somewhere sth. IMHO starting point should be "instead of asking when someone will do change you want to see, raise PR yourself - there is no need to ask maintainers why they don't deliver for you in timeline you want".

**Reply by Wirone** [reply]
> For stale bot, I believe that we can allow stale bot to close issues without your manual supervision. If person who was interested in topic will be still interested in it, they can always ask to reopen or... wait for it... create a PR.
> and if we have idea of long-term change, we park it under dedicated milestone.

It's more or less how I do it now, but issues that I already considered as outdated were closed during backlog review, some were closed after 2-3 Stale Bot iterations, some were moved to long term ideas milestone to exclude them from Stale Bot's umbrella. There still _are_ issues that I personally would close, but as far as I know you want them to be open (e.g. really old PHPUnit migration rules). There are also issues that I don't consider long-term, but still relevant (hard to explain that, basically it means I give myself time to postpone decision about the issue). In general, it's not a big thing for me if Stale Bot once in few months tries to close issue and we decide to keep it, that's what maintenance is for me too. At some point decision may be changed or there will be time for implementation.
